### ISSUE TITLE
Do not archive ads and trackers

### ISSUE THREAD

**Issue by BrandonOCasey** [pirate/ArchiveBox#211]
## Type 

 - [ ] General Question or Disussion
 - [x] Propose a brand new feature
 - [ ] Request modification of existing behavior or design

## What is the problem that your feature request solves
Archive pages without ads or trackers

## Describe the ideal specific solution you'd want, and whether it fits into any broader scope of changes
A flag that defaults to off for archiving without ads/trackers.

## What hacks or alternative solutions have you tried to solve the problem?
None 

## How badly do you want this new feature?

 - [ ] It's an urgent deal-breaker, I cant live without it
 - [x] It's important to add it in the near-mid term future
 - [ ] It would be nice to have eventually

---

 - [x] I'm willing to contribute to development / fixing this issue
 - [x] I like ArchiveBox so far / would recommend it to a friend

**Comment by wintdkyo** [comment]
One potential solution outside of ArchieveBox would be to have some type of a network-wide DNS ad blocking, such as a Pi-hole

**Comment by knowncolor** [comment]
I haven't tried this, but you might be able to achieve similar by installing certain extensions uBlock Origin, Decentraleyes, etc.

**Comment by pirate** [comment]
Currently, adblocking during PDF/Screenshot/DOM archiving is already possible by using a `CHROME_USER_DATA_DIR` with a profile that has an adblocker installed.  In the future it will be doable in a few more ways though:

1. Using a proxy for all archiving that removes ads e.g. `chromium --proxy-server=http://localhost:8080 --ignore-certificate-errors --disable-web-security https://some.site.withads.com`
2. Using a DNS resolver that blocks ads (e.g. Pi-hole)
3. Using a chrome extension like Ublock Origin and/or Ghostery
4. Using user scrips loaded during archiving to manually remove ads from specific sites

As we get closer to integrating pyppeteer (#177) and pywb (#130) I'll check back and update this issue with progress.

**Comment by onemenzel** [comment]
Has anyone tried installing extensions into the chrome profile? I'm running on docker compose and during my research journey on how to install extensions into a headless chrome from the shell, I found this chromium thread saying that there is no extension support in headless: https://bugs.chromium.org/p/chromium/issues/detail?id=706008

**Comment by pirate** [comment]
You're correct that extensions only work in non-headless mode, but at least playwright makes it easy https://playwright.dev/docs/chrome-extensions. One way around that might be to use a fake display and attach chrome to it. Then we could even make an http endpoint where you can watch that display to see a realtime view of what the worker is seeing while archiving.

**Comment by onemenzel** [comment]
> You're correct that extensions only work in non-headless mode, but at least playwright makes it easy https://playwright.dev/docs/chrome-extensions. One way around that might be to use a fake display and attach chrome to it. Then we could even make an http endpoint where you can watch that display to see a realtime view of what the worker is seeing while archiving.
> 
> You can also load extension sources as ordinary JS scripts in context of the page in playwright, instead of using the real extensions system. You just need the original source of the extension, which could be ripped out of the published extensions in minified form, or copied from github if they're open-source.

I think that'd be interesting features for future versions if some configuration method is exposed to the users. For now, I resigned to using dnsforge.de as the DNS server insode the container. They do not log DNS queries and block ad domains. Not quite as good as uBlock origin but it'll do it for now.

For everyone who's searching for a method to do this in docker-compose: it's as easy as adding the line `dns: your.dns.servers.ip` to the service in the `docker-compose.yml` file.

Thanks for your time and that nice piece of software!

**Comment by akhilleusuggo** [comment]
I'm using pihole as my main DNS ( on rpi ), and works perfectly. No ads or trackers during archiving.
Could also be installed as docker, on the same network bridged and point to it.

**Comment by nicoandmee** [comment]
> Actually pihole is a great idea, we should add a commented out example container in `docker-compose.yml` that shows how to setup pihole with the archivebox container's DNS pointed to the pihole container. That would be a very practical solution that doesn't involve a lot of dev work to add scripting during archiving.
> 
> I've just added it to the `docker-compose.yml` file: [5cd2b32](https://github.com/ArchiveBox/ArchiveBox/commit/5cd2b328c006db9e39b919a8d7ab01aaecee2fe9)

I may be being dumb, but shouldn't the port mapping be different for pihole? As in, not `80:80`, since port 80 is already being used by the principal web UI?

**Comment by pirate** [comment]
Port 80 is not used by the principal UI.

**Comment by nicoandmee** [comment]
> Port 80 is not used by the principal UI by default.

Yeah, my mistake- I'm not sure why but I could not get pihole to run as part of docker-compose, so I just set it up outside on my LAN. In case someone else runs into issues, be careful putting `network_mode: host` in the pihole compose config. I did so without completely understanding what it does- and it totally screwed up my network requiring a factory reset of my wifi points. Don't ask me why... 😓

**Comment by akhilleusuggo** [comment]
Not sure how can a docker-compose break a WIFI... I'm not gonna ask why.

But what you have to do, is add an entry on the compose ( on the archivebox section ), with DNS pointing at pihole.

In my case I'm running them separately, but I'll test them and leave an example.

**Comment by pirate** [comment]
I've already added an example `pihole` `docker-compose.yml` config here you can use: https://github.com/ArchiveBox/ArchiveBox/blob/dev/docker-compose.yml#L49

**Comment by akhilleusuggo** [comment]
@pirate Yes I've seen that, but maybe I'm missunderstanding something.

How does archivebox knows that PiHole Local IP is a DNS server? What I'm trying to say is that, requires extra environments variables to point to Pihole IP as DNS server. Under section of `archivebox`, something like this should be added;

```
dns:
      - 172.21.0.2
```

And Pihole should be configured with a static IP; the same one under the `dns` section.

**Comment by slikas** [comment]
Linking chrome profile with extension ublock-origin doesn't work. (I need to remove cookie consent panels and HTML elements, AdGuardHome doesn't block cookie panels and leaves big blank ad spaces)

I copied the profile dir into dockerized archivebox /data
Then set in `ArchiveBox.conf`:
`CHROME_USER_DATA_DIR=/data/chrome`
`CHROME_HEADLESS=False`
Then `docker run` it, no issues reported

But as said, no adblock applied to saved contents

**Comment by diego898** [comment]
Just pinging this issue again!

**Comment by pirate** [comment]
Quick update: The recommended solution is still to use pihole in front of ArchiveBox for now, I cant promise browser-extension-based ArchiveBox adblocking will be implemented anytime soon.

**Comment by bogorad** [comment]
What I tried:

1. Migrated a working profile
1.1 created a Chrome profile in an LXC container
1.2 installed and set up the extensions required:
1.2.1 [uBlock Origin](https://chrome.google.com/webstore/detail/ublock-origin/cjpalhdlnbpafiamejdnhcphjbkeiagm)
1.2.2 [I Still Don't Care About Cookies](https://chrome.google.com/webstore/detail/i-still-dont-care-about-c/edibdbjcniadpccecjdfdjjppcpchdlm)
1.2.3 [Bypass Paywalls Clean](https://gitlab.com/magnolia1234/bypass-paywalls-chrome-clean)
1.3 zipped the profile directory
1.4 transferred it into my Archivebox container
1.5 unpacked it into `/data/chrome-profile`
1.6 set up the correct ownership (chown -R ...)
1.7 added env-vars to my docker-compose:
1.7.1 `CHROME_USER_DATA_DIR = /data/chrome-profile`
1.7.2 `CHROME_HEADLESS = False`

When I tried adding pages to Archivebox via UI, it started failing, logging an error (console):

```
*** stack smashing detected ***: terminated
```

2. Then I tried running chromium to see what's what. 
2.1 did a `docker exec -ti archivebox cat /etc/passwd|grep archivebox` to discover the `userid`
2.2 did a `docker exec -ti --user=<userid> archivebox bash`
2.3 inside I set up an X-server address `(DISPLAY=<IP>:<id>`
2.4 ran chromium: `chromium-browser --no-sandbox --user-data-dir=/data/chrome-profile/`

All my extensions from the imported profile show fine, but the error (stack smashing) persists.

I also tried completely removing the imported profile so that chromium would create a new one. This way, chromium starts, but fails on the chrome-web-store page that installs uBlock origin (and many other pages).

Is there a better way to import a profile? Or am I doing something wrong?

**Comment by pirate** [comment]
Using external dns is very easy @zero77, no need for pihole, you just set the lines in the docker-compose.yml to point to any server you want:

```yml
services:
  archivebox:
    ...
    dns:
      - 1.1.1.1
      - ... etc any ip can go here ...
```

**Comment by turian** [comment]
Just curious the progress on this. What is the current best practice and will something more easy to set up / native be added to archivebox

**Comment by pirate** [comment]
Current best practice that I run myself is:

- install uBlock origin in a chrome profile + [set up archivebox to use that profile](https://github.com/ArchiveBox/ArchiveBox/wiki/Chromium-Install#setting-up-a-chromium-user-profile) (this prevents 99% of ads)
- set up pihole as indicated in `docker-compose.yml` (this gets the last 1%)
- configure `URL_DENYLIST` to prevent archiving of certain URLs/patterns (this covers other common minor nits/annoyance cases where you might want to avoid archiving some unecessary content)

This is as "native" as it gets right now. I have plans to make the chrome profile setup process easier (e.g. ability to specify extensions to install automatically), and I think adblocking is just one of the things that will become easier as a result.
